src/app/modules/auth.rs

```diff
@@ -1,18 +1,246 @@
+    fn pub_key(&self) -> &Self::PubKey {
+        unimplemented!()
+    }
+
+    fn number(&self) -> u64 {
+        self.number
+    }
+
+    fn sequence(&self) -> u64 {
+        self.sequence
+    }
+}
+
+impl Protobuf<BaseAccount> for AuthAccount {}
+
+impl TryFrom<BaseAccount> for AuthAccount {
+    type Error = String;
+
+    fn try_from(account: BaseAccount) -> Result<Self, Self::Error> {
+        Ok(AuthAccount {
+            address: account
+                .address
+                .parse()
+                .map_err(|_| "Failed to parse address".to_string())?,
+            number: account.account_number,
+            sequence: account.sequence,
+        })
+    }
+}
+
+impl From<AuthAccount> for BaseAccount {
+    fn from(account: AuthAccount) -> Self {
+        BaseAccount {
+            address: account.address.to_string(),
+            pub_key: None,
+            account_number: account.number,
+            sequence: account.sequence,
+        }
+    }
+}
+
+impl From<AuthAccount> for Any {
+    fn from(account: AuthAccount) -> Self {
+        let account = BaseAccount::from(account);
+        Any {
+            type_url: "/cosmos.auth.v1beta1.BaseAccount".to_string(),
+            value: account.encode_to_vec(),
+        }
+    }
+}
+
+pub trait AccountReader {
+    type Error;
+    type Address;
+    type Account: Account;
+
+    fn get_account(&self, address: Self::Address) -> Result<Self::Account, Self::Error>;
+}
+
+pub trait AccountKeeper {
+    type Error;
+    type Account: Account;
+
+    fn set_account(&mut self, account: Self::Account) -> Result<(), Self::Error>;
+
+    fn remove_account(&mut self, account: Self::Account) -> Result<(), Self::Error>;
+}
+
+#[derive(Clone)]
+pub struct Auth<S> {
+    store: SharedStore<S>,
+    account_reader: AuthAccountReader<S>,
+    account_keeper: AuthAccountKeeper<S>,
+}
+
+impl<S: 'static + ProvableStore> Auth<S> {
+    pub fn new(store: SharedStore<S>) -> Self {
+        Self {
+            store: store.clone(),
+            account_reader: AuthAccountReader {
+                account_store: TypedStore::new(store.clone()),
+            },
+            account_keeper: AuthAccountKeeper {
+                account_store: TypedStore::new(store),
+            },
+        }
+    }
+
+    pub fn query(&self) -> QueryServer<AuthQuery<S>> {
+        QueryServer::new(AuthQuery {
+            account_reader: self.account_reader().clone(),
+        })
+    }
+
+    pub fn account_reader(&self) -> &AuthAccountReader<S> {
+        &self.account_reader
+    }
+
+    pub fn account_keeper(&self) -> &AuthAccountKeeper<S> {
+        &self.account_keeper
+    }
+}
+
+impl<S: Store> Module<S> for Auth<S> {
+    fn init(&mut self, app_state: Value) {
+        debug!("Initializing auth module");
+        // safety - we panic on errors to prevent chain creation with invalid genesis config
+        let accounts: HashMap<String, HashMap<Denom, u64>> =
+            serde_json::from_value(app_state).unwrap();
```

Comment:
Why not parse `app_state` into `accounts: HashMap<>` in `init_chain()`, and pass `accounts` in `init()` instead? Currently, the `bank` and `auth` modules both need to parse the json separately.

It's also better documentation for what the genesis file is supposed to look like.

Reply:
Thinking about this more, I feel it's better to let modules decide how they want to use the genesis `app_state`. This is because modules should be able to decide what data they're interested in and how they want to interpret it. WDYT?

Reply:
That is not something modules can decide, right? Each module is given the `app_state` in whatever form, and from there each module can use whichever part they care about. Right now, we use `json::Value`, but really, there's an implicit structure that `json::Value` doesn't capture. Concretely, all modules **will have to** parse it as
```rust
let accounts: HashMap<String, HashMap<Denom, u64>> =
    serde_json::from_value(app_state).unwrap();
```
otherwise the `from_value()` will fail. My point is: why not make that explicit in the type?

We can deal with that later though I don't think it's a priority right now.